proposals/p2687.md

```diff
@@ -0,0 +1,298 @@
+error even though it would in fact complete if allowed to continue running. We
+would like a criteria that correctly classifies the examples that arise in
+practice.
+
+## Background
+
+The first termination rule was introduced in proposal
+[#920: Generic parameterized impls (details 5)](https://github.com/carbon-language/carbon-lang/pull/920),
+following Rust and C++. The problems with using a recursion limit were
+[recognized at the time that proposal was written](https://github.com/carbon-language/carbon-lang/blob/f282bca20e41e2f8dc05881d9d6b38213d6c6c87/docs/design/generics/details.md#termination-rule),
+but no alternative was known.
+
+Alternatives termination rules have since been discussed:
+
+-   in open discussion on
+    [2022-04-13](https://docs.google.com/document/d/1tEt4iM6vfcY0O0DG0uOEMIbaXcZXlNREc2ChNiEtn_w/edit#heading=h.cja3fkwzv9tr),
+    prompted by a question on
+    [#1088: Generic details 10: interface-implemented requirements](https://github.com/carbon-language/carbon-lang/pull/1088);
+    and
+-   in issue
+    [#2458: Infinite recursion during impl selection](https://github.com/carbon-language/carbon-lang/issues/2458),
+    which includes summaries of discussions including those on
+    [2023-02-07](https://docs.google.com/document/d/1gnJBTfY81fZYvI_QXjwKk1uQHYBNHGqRLI2BS_cYYNQ/edit?resourcekey=0-ql1Q1WvTcDvhycf8LbA9DQ#heading=h.9u2u6078figt).
+
+PR
+[#2602: Implement the termination algorithm for impl selection described in #2458](https://github.com/carbon-language/carbon-lang/pull/2602)
+implements the termination rule of this proposal in Explorer.
+
+## Proposal
+
+We replace the termination criteria with a rule that the types in the `impl`
+query must never get strictly more complicated. The way we measure the
+complexity of a set of types is by counting how many of each base type appears.
+A base type is the name of a type without its parameters. For example, the base
+types in this query `Pair(Optional(i32), bool) impls AddWith(Optional(i32))`
+are:
+
+-   `Pair`
+-   `Optional` twice
+-   `i32` twice
+-   `bool`
+-   `AddWith`
+
+A query is strictly more complicated if at least one count increases, and no
+count decreases. So `Optional(Optional(i32))` is strictly more complicated than
+`Optional(i32)` but not strictly more complicated than `Optional(bool)`.
+
+This rule, when combined with
+[the acyclic rule](/docs/design/generics/details.md#acyclic-rule) that a query
+can't repeat exactly, [guarantees termination](#proof-of-termination). This rule
+is expected to identify a problematic sequence of `impl` declaration
+instantiations in a way that is easier for the user to understand. Consider the
+example from before,
+
+```
+interface I;
+impl forall [T:! type where Optional(.Self) is I] T as I;
+```
+
+This `impl` declaration matches the query `i32 impls I` as long as
+`Optional(i32) impls I`. That is a strictly more complicated query, though,
```

Comment:
One minor detail here: we should only report an error if we cycle back to considering the same `impl` declaration again. I think we wanted the type structure ordering / `match_first` ordering to apply before we detect cycles, so that we accept:

```
impl forall [T:! type where Optional(.Self) is I] T as I;
impl Optional(i32) as I;
// OK, because we never consider the first `impl` when looking for `Optional(i32) impls I`.
let T:! I = i32;
// error, cycle
let U:! I = f32;
```

Reply:
Added some text above and below.